File: LeetCode/647.js
/**
 * @param {string} s
 * @return {number}
 */
var countSubstrings = function(s) {
    let count = 0;
    const len = s.length;

    for (let i = 0; i < 2 * len - 1; ++i) {
        const l = Math.floor(i / 2);
        const r = Math.ceil(i / 2);
        let j = 0;
        while (l - j >= 0 && r + j < len && s[l - j] === s[r + j]) ++j;
        count += j;
    }

    return count;
};

File: LeetCode/647.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int count = 0;
        const int len = s.length();

        for (int i = 0; i < 2 * len - 1; ++i) {
            int l = i / 2;
            int r = i / 2 + i % 2;
            while(l >= 0 && r < len && s[l] == s[r]) {
                ++count;

                --l;
                ++r;
            }
        }

        return count;
    }
};

class Solution {
public:
    int countSubstrings(string s) {
        int cnt = 0;
        vector<vector<bool>> dp(s.length() + 1, vector<bool>(s.length() + 1, false));

        for (int i = s.length(); i >= 1; i--) {
            for (int j = i; j <= s.length(); j++) {
                if (s[i - 1] == s[j - 1] && (j - i == 0 || j - i == 1 || dp[i + 1][j - 1])) {
                    ++cnt;
                    dp[i][j] = true;
                }
            }
        }

        return cnt;
    }
};
